### PR TITLE
Metricsregistry

### LINKED ISSUE #63: Handling of conflicting metric values
We had some developers on a new project starting to integrate statsd support in their application. `statsd_exporter` 0.3.0 is crashing constantly with this error message:

```
FATA[0000] A change of configuration created inconsistent metrics for "query_timer". You have to restart the statsd_exporter, and you should consider the effects on your monitoring setup. Error: a previously registered descriptor with the same fully-qualified name as Desc{fqName: "query_timer", help: "Metric autogenerated by statsd_exporter.", constLabels: {after_date="2017-01-19 00:00:00-05:00",application="…",component="…",environment="testing",server="…",sub_component="…"}, variableLabels: []} has different label names or a different help string  source=exporter.go:137
```

If I'm reading that correctly, this due to them having two versions of the app sending inconsistent metric values but while an error in this situation seems reasonable it actually causes the `statsd_exporter` process to crash.

### PR DESCRIPTION
**Pull request by jacksontj**
Rebase of #74, fixes #63 

**TLDR; statsd exporter currently doesn't allow for duplicate/conflicting tag-sets -- which is a problem for the statsd exporter. The exporter interface has no such restriction, its simply an issue with the go prometheus client. This can be fixed by implementing our own registry-- which is this PR.**

Apologies for the long delay, other things came up as higher priority. So here is a rebased version against current master to start the conversation from.

As far as the outstanding comments from the previous PR:

`Please update the help string for both flags to make the difference very very clear.` -- Updated with the most recent commit on the branch

`What happens when these are exceeded? What happens for someone with very many metrics?` -- this is a copy/pasted hard-coded value from the [prometheus client_golang](https://github.com/prometheus/client_golang/blob/master/prometheus/registry.go#L31) when the channel hits capacity it will block collection until it has been consumed (effectively a memory limiter).

`that PR is closed. what's the future here?` -- We'll have to maintain this registry long-term. If we'd prefer I can pull the registry into a separate package (directory in the same repo).

cc: @matthiasr

### CODE REVIEW

File: registry.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+
+	// Scatter.
+	// (Collectors could be complex and slow, so we call them all at once.)
+	wg.Add(len(r.metricsByID))
+	go func() {
+		wg.Wait()
+		close(metricChan)
+	}()
+	for _, metric := range r.metricsByID {
+		go func(collector prometheus.Collector) {
```
Comment (matthiasr) [inline registry.go]: As pointed out in https://github.com/prometheus/statsd_exporter/issues/76#issuecomment-359783877 this causes a huge spike in memory consumption because it spawns a goroutine for every single time series.

prometheus/client_golang#370 is a potential fix for that.

**Comment by jacksontj** [comment]
If we end up needing to continue with the registry we can either optimize
or implement a more naive one. I'd be more than happy to not "build our
own" but the response last time this was discussed upstream was basically
"try it yourself and of we like it we might merge it".

**Comment by beorn7** [comment]
Re-implementing the registry to create inconsistent metrics sounds like a really bad idea, to be honest.
The ideas for https://github.com/prometheus/client_golang/issues/47 and https://github.com/prometheus/client_golang/issues/355 are around allowing collectors to be declared "unchecked", and then the registry in the client library will try its best to pad missing labels (using empty label values). This will guarantee that every single exposition is consistent, but it will not guarantee consistency over the lifetime of a binary (which is probably a price we can/need to pay to get out of this coundrum).

**Comment by matthiasr** [comment]
So we just discussed this separately – the linked issues don't quite cut it for the statsd exporter. They work well with custom collectors that generate ConstMetrics, but this exporter doesn't. It could somewhat sensibly for counters and gauges, but we'd have to recreate the quantile calculation from scratch for summaries, and the bucketing for histograms.

Another idea that we've kicked around is tracking the labels for a given metric name (this would be easy on top of #119), and when they change unregister the metric and register a completely new one. That's not allowed at the moment anyway, but could be without breaking the semantics too much. On top of that, we could fill out with empty labels as needed anyway. This would mean that at the transition, we would throw away all previous observations (unless we dig into the protobuf to rescue them and copy them over into the new metrics somehow.)

There's no really pretty solution yet …

**Comment by beorn7** [comment]
More thoughts:

The "unregister" hack could be implemented more cleanly (and without changing the semantics of `Unregister`) by using a local (not the global) registry and simply register all metrics (the newly labeled and the old ones that haven't changed) with the new registry. Then throw the old HTTP handler and the old registry away and wire up the new registry.

WRT throwing away the old observations: With summaries, you want a decay time anyway (by default 10m), so assuming label changes are rare, that's not a big deal. For histogram, throwing away the old observations is essentially a counter reset and will be dealt with by Prometheus, i.e. you'll only lose about half a scrape interval worth of even counts.

I'd say let's go with the "throw away on label change" approach, i.e. whenever a previously unaccounted label shows up, recreate all affected metrics with a zero count in a new registry, register all unaffected metrics with the new registry, rewire the HTTP handle to the new registry and throw away the old one.

Does that make sense?

**Comment by matthiasr** [comment]
Could there be any strange side effects from not resetting all metrics at
the same time? Say, for request totals and errors?

**Comment by brian-brazil** [comment]
Technically no, but I can imagine that causing artifacts. On the other hand applications should not be changing labels at runtime.

**Comment by matthiasr** [comment]
The statsd exporter is not, by itself, an application – I would still recommend coupling its lifecycle to that of the application but I also recognize that that is _not_ making it a drop-in replacement for statsd (where a central deployment is much more common). Now, with the schema @beorn7 proposed, we would re-register and reset the metric that changed, but we would not have a mechanism to do that to _all_ metrics exposed by one application at the same time (because we don't know what these all are).

I'm not saying it's a bad idea, just writing down thoughts about the limitations. Whatever we choose to do, I'd want to document the different choices and tradeoffs really well. We could even make the metrics-reregistering an optional feature (and drop the event otherwise).

**Comment by beorn7** [comment]
> Could there be any strange side effects from not resetting all metrics at the same time? Say, for request totals and errors?

To quote from my Prometheus proverb collection: “First do `rate`, then aggregate.” That means, that counter resets are taken care of before aggregating. I guess the scenario you have in mind is a counter for total, that doesn't change its labels, while suddenly error counters arrive with an additional label. Thus, the error counters get a reset but not the total counter. With the usual exception of losing a few increments, it should be fine.

**Comment by matthiasr** [comment]
@jacksontj is this (wiping the whole metric and recreating with the new label set) a solution that would fit your needs?

**Comment by grobie** [comment]
How would that work out during a deployment, where some instances send the new label set and others still the old one? If I understand the proposal correctly, the exporter would constantly throw away the existing metric and create a new one.

**Comment by matthiasr** [comment]
I wouldn't do this for shrinking label sets, those I would just pad out indefinitely. So it would only happen once per metric name, the first time an event with the wider label set is seen. Renaming a label would result in a "wider" label set that includes both.

**Comment by jacksontj** [comment]
@matthiasr that would mean that the metrics sent from clients to this exporter would have labels *added* with empty values. IMO thats not workable, as that means I can no longer query based on the existence of label keys -- since they are being mutated before storage. 

the intention of this exporter is to be (as much as possible) a drop-in-replacment for a statsd server/aggregator/relay/etc. With that in mind we should avoid mutating the metrics as much as possible-- we want to store exactly what we where sent.

**Comment by matthiasr** [comment]
Fundamentally, that's an impossible task. The data models are not the same so we can't store exactly what we get sent – only Graphite can.

The Prometheus server does not distinguish between empty and non-present labels, so you can't query based on label key presence anyway – the query language does not allow that. The only way to test for "label is not present" is to match `{label=""}` anyway.

**Comment by matthiasr** [comment]
I'm going to close this – the concrete changeset is very old, and I believe with prometheus/client_golang#425 we can solve this without a whole new registry.

The concrete problem in #63 was fixed from another angle.
